Refactor Todo List App

mrkeramatiyan opened the pull request:
-be able to create to multiple todo
-set optional preset for todo
-Add button for insert task
-change ui

mrkeramatiyan committed:
Refactor Todo List App

-be able to create to multiple todo
-set optional preset for todo
-Add button for insert task
-change ui

orkideh commented:
I appreciate the short bullet points, I think it will be even a bit better if you would choose more descriptive name for your PR, branch and commits, Also a link to result to review the changes.

orkideh commented:
The app UI looks quite appealing and that is very brilliant idea to have input validation to check duplicates, however I think you can still improve it by: 
- Communicating to user by a simple [Snackbar / Toast](https://www.w3schools.com/howto/howto_js_snackbar.asp) rather than "alert" as it is UI blocking.
- Also I have noticed if there is already an item "AAA" exist in the list user can create another item for example "BBB" then rename it to "AAA" and if user remove any of those both will be removed.  I assume is not desired behavior of your app.
- I would suggest to find a way to allow user to have any items with any name(event duplicated) and you handle any item separated not matter of the title.

orkideh commented on todo-list/script.js:
```diff
@@ -1,150 +1,231 @@
+class TodoList {
+    
+    constructor(elementInitApp, preset = {}) {
+        this.titleOfApp = preset.title ? preset.title : 'Todo List App';
```
You may consider using [default operator](https://topherlandry.wordpress.com/2014/09/06/the-javascript-guard-and-default-operators/) in JS. like:
```
this.titleOfApp = preset.title || 'Todo List App';
```

orkideh commented on todo-list/script.js:
```diff
@@ -1,150 +1,231 @@
+    
+    constructor(elementInitApp, preset = {}) {
+        this.titleOfApp = preset.title ? preset.title : 'Todo List App';
+        this.idOfApp = preset.id ? preset.id : '';
```
same "default operator"

orkideh commented on todo-list/script.js:
```diff
@@ -1,150 +1,231 @@
-            todos = todos.filter(todoObj => todoObj.label != todos[id].label);
-            renderTodos();
+    checkDataForState() {
+        const getDataFromlocalStorage = localStorage.getItem(this.elementInitApp.getAttribute('id'));
```
This called hydration technique, big kudos.

orkideh commented on todo-list/script.js:
```diff
@@ -1,150 +1,231 @@
+        const element =  document.createElement(elementName);
+        if(addClass) element.classList.add(addClass);
+        if(value) element.innerText = value;
+        if(id) element.setAttribute('id', id + '');
```
Why `id + '' `

mrkeramatiyan replied:
because sometimes not set id to element and i think it should again convert to string

orkideh commented on todo-list/script.js:
```diff
@@ -1,150 +1,231 @@
+                value = this.closest('section').querySelector('.input-text-task').value;         
+            }
+            
+            const result = Boolean(mainContext.state.find(taskObj => taskObj.label == value));
```
You may also use [Array.some](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some)

orkideh commented on todo-list/script.js:
```diff
@@ -1,150 +1,231 @@
+            if(result) {
+                alert('Your task is already have!!!')
+            }
+            if(value != '' && value != ' ' && !result) {
```
You may "trim" the `value` before or use `value.trim()` to remove all the unnecessary white spaces in your condition.

mrkeramatiyan replied:
you are right, i used trim() in lined 66 but i forgot use in line 68

orkideh commented on todo-list/script.js:
```diff
@@ -1,150 +1,231 @@
+            const guid = this.genarateUID(5);
+
+            const li = this.createElement('li');
+            li.setAttribute('id', `${index}`);
```
Why `${index}` ?

mrkeramatiyan replied:
i setted index of array to each li for find exact data

orkideh commented on todo-list/script.js:
```diff
@@ -1,150 +1,231 @@
+    }
+
+    saveToLocalStorage() {
+        const nameDataFromId = this.idOfApp ? this.idOfApp : this.elementInitApp.getAttribute('id');
```
same default value

mrkeramatiyan committed:
some change & refactor

- Add toaster for alert and confrim
- I used your suggestions in code like Array.some and default operator
- I refactored my todolist for duplicat task

mrkeramatiyan commented:
Preview link:

orkideh commented on todo-list/script.js:
```diff
@@ -1,150 +1,314 @@
+class TodoList {
+    
+    constructor(elementInitApp, preset = {}) {
+        this.titleOfApp = preset.title || 'Todo List App';
```
You may consider using [Object.assing](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign) to assign default params

orkideh commented on todo-list/script.js:
```diff
@@ -1,150 +1,314 @@
+        // function of event
+        function handlerAddTaskBtn(event) {
+            let value = '';
+            if(this.tagName == 'INPUT') {
```
I suggest to use  [Event.target](https://developer.mozilla.org/en-US/docs/Web/API/Event/target) or [Event.currentTarget](https://developer.mozilla.org/en-US/docs/Web/API/Event/currentTarget) rather than the  context `this`

mrkeramatiyan replied:
thanks for the links, they have been really informative, but yet I do not understand the benefit of them rather than context `this`

orkideh commented on todo-list/script.js:
```diff
@@ -1,150 +1,314 @@
-        todos[id].checked = !todos[id].checked;
-        renderTodos();
+    renderTodoTasks() {
+        const fragment = document.createDocumentFragment();
```
👍

mrkeramatiyan committed:
Changes

- changed with your suggestions
